File: ios/FT8AFKit/Sources/FT8Engine/WsjtxCodec.swift
import Foundation

/// Byte-exact (de)serialization of the WSJT-X UDP message protocol.
///
/// WSJT-X broadcasts a documented binary protocol over UDP that GridTracker,
/// JTAlert, N1MM+, Log4OM etc. consume, and accepts a few request messages so
/// those apps can drive it. This speaks that exact wire format so FT8AF is a
/// drop-in source/sink.
///
/// The framing is a bespoke big-endian Qt-`QDataStream` layout (NOT JSON), so
/// byte-for-byte correctness is the whole game — every encoder here is pinned by
/// the same golden vectors as the desktop (Rust) and Android (Kotlin) codecs
/// (see `docs/wsjtx-udp.md` and `WsjtxCodecTests`). Pure, no I/O —
/// `WsjtxUdpService` owns the socket.
///
/// Wire primitives (all big-endian): u32/u64/i32/i64/f64, bool(1 byte); string =
/// u32 byte-length (0xFFFFFFFF = null) then UTF-8 bytes (empty = length 0);
/// datetime (UTC) = i64 Julian day, u32 ms-since-midnight, u8=1. Every datagram:
/// u32 magic 0xADBCCBDA, u32 schema, u32 type, string id, then type fields.
public enum WsjtxCodec {
    public static let magic: UInt32 = 0xADBC_CBDA
    public static let schema: UInt32 = 3
    public static let clientId = "FT8AF"

    // Message type numbers.
    public static let tHeartbeat: UInt32 = 0
    public static let tStatus: UInt32 = 1
    public static let tDecode: UInt32 = 2
    public static let tClear: UInt32 = 3
    public static let tReply: UInt32 = 4
    public static let tQsoLogged: UInt32 = 5
    public static let tReplay: UInt32 = 7
    public static let tHaltTx: UInt32 = 8
    public static let tFreeText: UInt32 = 9
    public static let tLoggedAdif: UInt32 = 12

    // MARK: - outbound field carriers

    public struct Status {
        public var dialFreqHz: UInt64
        public var mode: String
        public var dxCall: String
        public var report: String
        public var txMode: String
        public var txEnabled: Bool
        public var transmitting: Bool
        public var decoding: Bool
        public var rxDf: UInt32
        public var txDf: UInt32
        public var deCall: String
        public var deGrid: String
        public var dxGrid: String
        public var txWatchdog: Bool
        public var subMode: String
        public var fastMode: Bool
        public var specialOpMode: UInt8
        public var freqTolerance: UInt32
        public var trPeriod: UInt32
        public var configName: String
        public var txMessage: String

        public init(
            dialFreqHz: UInt64, mode: String, dxCall: String, report: String,
            txMode: String, txEnabled: Bool, transmitting: Bool, decoding: Bool,
            rxDf: UInt32, txDf: UInt32, deCall: String, deGrid: String, dxGrid: String,
            txWatchdog: Bool, subMode: String, fastMode: Bool, specialOpMode: UInt8,
            freqTolerance: UInt32, trPeriod: UInt32, configName: String, txMessage: String
        ) {
            self.dialFreqHz = dialFreqHz; self.mode = mode; self.dxCall = dxCall
            self.report = report; self.txMode = txMode; self.txEnabled = txEnabled
            self.transmitting = transmitting; self.decoding = decoding; self.rxDf = rxDf
            self.txDf = txDf; self.deCall = deCall; self.deGrid = deGrid; self.dxGrid = dxGrid
            self.txWatchdog = txWatchdog; self.subMode = subMode; self.fastMode = fastMode
            self.specialOpMode = specialOpMode; self.freqTolerance = freqTolerance
            self.trPeriod = trPeriod; self.configName = configName; self.txMessage = txMessage
        }
    }

    public struct Decode {
        public var isNew: Bool
        public var timeMs: UInt32
        public var snr: Int32
        public var deltaTime: Double
        public var deltaFreq: UInt32
        public var mode: String
        public var message: String
        public var lowConfidence: Bool
        public var offAir: Bool

        public init(
            isNew: Bool, timeMs: UInt32, snr: Int32, deltaTime: Double, deltaFreq: UInt32,
            mode: String, message: String, lowConfidence: Bool, offAir: Bool
        ) {
            self.isNew = isNew; self.timeMs = timeMs; self.snr = snr
            self.deltaTime = deltaTime; self.deltaFreq = deltaFreq; self.mode = mode
            self.message = message; self.lowConfidence = lowConfidence; self.offAir = offAir
        }
    }

    /// A UTC instant as WSJT-X serializes it.
    public struct DateTime: Equatable {
        public var julianDay: Int64
        public var msOfDay: UInt32
        public init(julianDay: Int64, msOfDay: UInt32) {
            self.julianDay = julianDay; self.msOfDay = msOfDay
        }
        /// Build from ADIF `YYYYMMDD` + `HHMMSS` (UTC); zeroed on bad input.
        public static func fromAdif(_ dateYyyymmdd: String, _ timeHhmmss: String) -> DateTime {
            DateTime(
                julianDay: julianDayFromYyyymmdd(dateYyyymmdd) ?? 0,
                msOfDay: msOfDayFromHhmmss(timeHhmmss) ?? 0
            )
        }
    }

    public struct QsoLogged {
        public var timeOff: DateTime
        public var dxCall: String
        public var dxGrid: String
        public var txFreqHz: UInt64
        public var mode: String
        public var reportSent: String
        public var reportReceived: String
        public var txPower: String
        public var comments: String
        public var name: String
        public var timeOn: DateTime
        public var operatorCall: String
        public var myCall: String
        public var myGrid: String
        public var exchangeSent: String
        public var exchangeReceived: String
        public var propMode: String

        public init(
            timeOff: DateTime, dxCall: String, dxGrid: String, txFreqHz: UInt64, mode: String,
            reportSent: String, reportReceived: String, txPower: String, comments: String,
            name: String, timeOn: DateTime, operatorCall: String, myCall: String, myGrid: String,
            exchangeSent: String, exchangeReceived: String, propMode: String
        ) {
            self.timeOff = timeOff; self.dxCall = dxCall; self.dxGrid = dxGrid
            self.txFreqHz = txFreqHz; self.mode = mode; self.reportSent = reportSent
            self.reportReceived = reportReceived; self.txPower = txPower; self.comments = comments
            self.name = name; self.timeOn = timeOn; self.operatorCall = operatorCall
            self.myCall = myCall; self.myGrid = myGrid; self.exchangeSent = exchangeSent
            self.exchangeReceived = exchangeReceived; self.propMode = propMode
        }
    }

    // MARK: - writer

    private struct Writer {
        var data = Data()
        mutating func u8(_ v: UInt8) { data.append(v) }
        mutating func bool(_ v: Bool) { data.append(v ? 1 : 0) }
        mutating func u32(_ v: UInt32) { var b = v.bigEndian; withUnsafeBytes(of: &b) { data.append(contentsOf: $0) } }
        mutating func i32(_ v: Int32) { u32(UInt32(bitPattern: v)) }
        mutating func u64(_ v: UInt64) { var b = v.bigEndian; withUnsafeBytes(of: &b) { data.append(contentsOf: $0) } }
        mutating func i64(_ v: Int64) { u64(UInt64(bitPattern: v)) }
        mutating func f64(_ v: Double) { u64(v.bitPattern) }
        mutating func string(_ s: String) {
            let b = Array(s.utf8)
            u32(UInt32(b.count))
            data.append(contentsOf: b)
        }
        mutating func datetime(_ dt: DateTime) {
            i64(dt.julianDay)
            u32(dt.msOfDay)
            u8(1) // Qt::UTC
        }
    }

    private static func begin(_ type: UInt32) -> Writer {
        var w = Writer()
        w.u32(magic)
        w.u32(schema)
        w.u32(type)
        w.string(clientId)
        return w
    }

    // MARK: - outbound builders

    public static func heartbeat(version: String, revision: String) -> Data {
        var w = begin(tHeartbeat)
        w.u32(schema) // max schema
        w.string(version)
        w.string(revision)
        return w.data
    }

    public static func status(_ s: Status) -> Data {
        var w = begin(tStatus)
        w.u64(s.dialFreqHz)
        w.string(s.mode)
        w.string(s.dxCall)
        w.string(s.report)
        w.string(s.txMode)
        w.bool(s.txEnabled)
        w.bool(s.transmitting)
        w.bool(s.decoding)
        w.u32(s.rxDf)
        w.u32(s.txDf)
        w.string(s.deCall)
        w.string(s.deGrid)
        w.string(s.dxGrid)
        w.bool(s.txWatchdog)
        w.string(s.subMode)
        w.bool(s.fastMode)
        w.u8(s.specialOpMode)
        w.u32(s.freqTolerance)
        w.u32(s.trPeriod)
        w.string(s.configName)
        w.string(s.txMessage)
        return w.data
    }

    public static func decode(_ d: Decode) -> Data {
        var w = begin(tDecode)
        w.bool(d.isNew)
        w.u32(d.timeMs)
        w.i32(d.snr)
        w.f64(d.deltaTime)
        w.u32(d.deltaFreq)
        w.string(d.mode)
        w.string(d.message)
        w.bool(d.lowConfidence)
        w.bool(d.offAir)
        return w.data
    }

    public static func clear() -> Data { begin(tClear).data }

    public static func qsoLogged(_ q: QsoLogged) -> Data {
        var w = begin(tQsoLogged)
        w.datetime(q.timeOff)
        w.string(q.dxCall)
        w.string(q.dxGrid)
        w.u64(q.txFreqHz)
        w.string(q.mode)
        w.string(q.reportSent)
        w.string(q.reportReceived)
        w.string(q.txPower)
        w.string(q.comments)
        w.string(q.name)
        w.datetime(q.timeOn)
        w.string(q.operatorCall)
        w.string(q.myCall)
        w.string(q.myGrid)
        w.string(q.exchangeSent)
        w.string(q.exchangeReceived)
        w.string(q.propMode)
        return w.data
    }

    public static func loggedAdif(_ adif: String) -> Data {
        var w = begin(tLoggedAdif)
        w.string(adif)
        return w.data
    }

    // MARK: - inbound

    public enum Inbound: Equatable {
        case reply(call: String, grid: String, snr: Int32)
        case replay
        case haltTx(autoOnly: Bool)
        case freeText(text: String, send: Bool)
    }

    /// Bounds-checked big-endian reader; every read returns nil past the end.
    private struct Reader {
        let b: [UInt8]
        var pos = 0
        init(_ data: Data) { b = [UInt8](data) }
        mutating func take(_ n: Int) -> ArraySlice<UInt8>? {
            guard pos + n <= b.count else { return nil }
            let s = b[pos ..< pos + n]; pos += n; return s
        }
        mutating func u8() -> UInt8? { take(1)?.first }
        mutating func bool() -> Bool? { u8().map { $0 != 0 } }
        mutating func u32() -> UInt32? {
            guard let s = take(4) else { return nil }
            return s.reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
        }
        mutating func i32() -> Int32? { u32().map { Int32(bitPattern: $0) } }
        mutating func f64() -> Double? {
            guard let s = take(8) else { return nil }
            return Double(bitPattern: s.reduce(UInt64(0)) { ($0 << 8) | UInt64($1) })
        }
        mutating func string() -> String? {
            guard let len = u32() else { return nil }
            if len == 0xFFFF_FFFF { return "" }
            guard let s = take(Int(len)) else { return nil }
            return String(decoding: s, as: UTF8.self)
        }
    }

    /// Parse an inbound datagram into intent. Returns nil for anything we don't
    /// act on (bad magic, unknown/outbound-only type, truncated buffer).
    public static func parseInbound(_ data: Data) -> Inbound? {
        var r = Reader(data)
        guard r.u32() == magic else { return nil }
        guard r.u32() != nil else { return nil } // schema
        guard let type = r.u32() else { return nil }
        guard r.string() != nil else { return nil } // id
        switch type {
        case tReply:
            guard r.u32() != nil else { return nil }       // time
            guard let snr = r.i32() else { return nil }
            guard r.f64() != nil else { return nil }        // dt
            guard r.u32() != nil else { return nil }        // df
            guard r.string() != nil else { return nil }     // mode
            guard let message = r.string() else { return nil }
            guard let target = parseReplyTarget(message) else { return nil }
            return .reply(call: target.0, grid: target.1, snr: snr)
        case tReplay:
            return .replay
        case tHaltTx:
            return .haltTx(autoOnly: r.bool() ?? false)
        case tFreeText:
            guard let text = r.string() else { return nil }
            return .freeText(text: text, send: r.bool() ?? true)
        default:
            return nil
        }
    }

    /// From a decoded message line, work out which station a Reply should call and
    /// their grid. Handles `TO FROM ...` and CQ lines with an optional directive
    /// (`CQ FROM GRID`, `CQ DX FROM GRID`, `CQ POTA FROM GRID`).
    static func parseReplyTarget(_ message: String) -> (String, String)? {
        let tokens = message.split(whereSeparator: { $0 == " " || $0 == "\t" }).map(String.init)
        guard !tokens.isEmpty else { return nil }
        let callIdx: Int
        if tokens[0].uppercased() == "CQ" {
            guard tokens.count >= 2 else { return nil }
            callIdx = isCallsign(tokens[1]) ? 1 : 2
        } else {
            callIdx = 1
        }
        guard callIdx < tokens.count else { return nil }
        let call = tokens[callIdx]
        guard isCallsign(call) else { return nil }
        let grid = tokens[(callIdx + 1)...].first(where: isGrid4)?.uppercased() ?? ""
        return (call.uppercased(), grid)
    }

    private static func isCallsign(_ t: String) -> Bool {
        let core = t.split(separator: "/").first.map(String.init) ?? t
        let hasAlpha = core.contains { $0.isLetter }
        let hasDigit = core.contains { $0.isNumber }
        let okChars = t.allSatisfy { $0.isLetter || $0.isNumber || $0 == "/" }
        return hasAlpha && hasDigit && okChars && !isGrid4(t)
    }

    private static func isGrid4(_ t: String) -> Bool {
        let c = Array(t)
        guard c.count == 4 else { return false }
        func isAtoR(_ ch: Character) -> Bool {
            let u = Character(ch.uppercased()); return u >= "A" && u <= "R"
        }
        return isAtoR(c[0]) && isAtoR(c[1]) && c[2].isNumber && c[3].isNumber
    }

    // MARK: - date/time helpers

    /// Julian Day Number for a UTC `YYYYMMDD` date (matches Qt's toJulianDay).
    static func julianDayFromYyyymmdd(_ s: String) -> Int64? {
        guard s.count == 8, s.allSatisfy({ $0.isNumber }) else { return nil }
        let a = Array(s)
        guard let year = Int(String(a[0 ..< 4])),
              let month = Int(String(a[4 ..< 6])),
              let day = Int(String(a[6 ..< 8])),
              (1 ... 12).contains(month), (1 ... 31).contains(day) else { return nil }
        // Fliegel–Van Flandern JDN for a proleptic Gregorian date.
        let aa = (14 - month) / 12
        let y = year + 4800 - aa
        let m = month + 12 * aa - 3
        let jdn = day + (153 * m + 2) / 5 + 365 * y + y / 4 - y / 100 + y / 400 - 32045
        return Int64(jdn)
    }

    /// Milliseconds since midnight for a UTC `HHMMSS` string.
    static func msOfDayFromHhmmss(_ s: String) -> UInt32? {
        guard s.count == 6, s.allSatisfy({ $0.isNumber }) else { return nil }
        let a = Array(s)
        guard let h = Int(String(a[0 ..< 2])),
              let m = Int(String(a[2 ..< 4])),
              let sec = Int(String(a[4 ..< 6])),
              h <= 23, m <= 59, sec <= 60 else { return nil }
        return UInt32(((h * 3600) + (m * 60) + sec) * 1000)
    }
}

File: ft8af/app/src/main/kotlin/radio/ks3ckc/ft8af/wsjtx/WsjtxCodec.kt
package radio.ks3ckc.ft8af.wsjtx

import androidx.annotation.VisibleForTesting
import java.io.ByteArrayOutputStream
import java.io.DataOutputStream

/**
 * Byte-exact (de)serialization of the WSJT-X UDP message protocol.
 *
 * WSJT-X broadcasts a documented binary protocol over UDP that GridTracker,
 * JTAlert, N1MM+, Log4OM etc. consume, and accepts a few request messages so
 * those apps can drive it. This object speaks that exact wire format so FT8AF is
 * a drop-in source/sink.
 *
 * The framing is a bespoke big-endian Qt-`QDataStream` layout (NOT JSON), so
 * byte-for-byte correctness is the whole game — every encoder here is pinned by
 * the same golden vectors as the desktop (Rust) codec (see `docs/wsjtx-udp.md`
 * and `WsjtxCodecTest`). Pure, no I/O — [WsjtxUdpService] owns the socket.
 *
 * Wire primitives (all big-endian): u32/u64/i32/i64/f64, bool(1 byte); string =
 * u32 byte-length (0xFFFFFFFF = null) then UTF-8 bytes (empty = length 0);
 * datetime (UTC) = i64 Julian day, u32 ms-since-midnight, u8=1. Every datagram:
 * u32 magic 0xADBCCBDA, u32 schema, u32 type, string id, then type fields.
 */
object WsjtxCodec {
    @VisibleForTesting internal val MAGIC: Int = 0xADBCCBDA.toInt()
    const val SCHEMA: Int = 3
    const val CLIENT_ID: String = "FT8AF"

    // Message type numbers.
    const val T_HEARTBEAT = 0
    const val T_STATUS = 1
    const val T_DECODE = 2
    const val T_CLEAR = 3
    const val T_REPLY = 4
    const val T_QSO_LOGGED = 5
    const val T_REPLAY = 7
    const val T_HALT_TX = 8
    const val T_FREE_TEXT = 9
    const val T_LOGGED_ADIF = 12

    // --- outbound field carriers --------------------------------------------

    data class Status(
        val dialFreqHz: Long,
        val mode: String,
        val dxCall: String,
        val report: String,
        val txMode: String,
        val txEnabled: Boolean,
        val transmitting: Boolean,
        val decoding: Boolean,
        val rxDf: Int,
        val txDf: Int,
        val deCall: String,
        val deGrid: String,
        val dxGrid: String,
        val txWatchdog: Boolean,
        val subMode: String,
        val fastMode: Boolean,
        val specialOpMode: Int,
        val freqTolerance: Int,
        val trPeriod: Int,
        val configName: String,
        val txMessage: String,
    )

    data class Decode(
        val isNew: Boolean,
        val timeMs: Int,
        val snr: Int,
        val deltaTime: Double,
        val deltaFreq: Int,
        val mode: String,
        val message: String,
        val lowConfidence: Boolean,
        val offAir: Boolean,
    )

    /** A UTC instant as WSJT-X serializes it. */
    data class DateTime(val julianDay: Long, val msOfDay: Int) {
        companion object {
            /** Build from ADIF `YYYYMMDD` + `HHMMSS` (UTC); zeroed on bad input. */
            @JvmStatic
            fun fromAdif(dateYyyymmdd: String?, timeHhmmss: String?): DateTime =
                DateTime(
                    julianDayFromYyyymmdd(dateYyyymmdd) ?: 0L,
                    msOfDayFromHhmmss(timeHhmmss) ?: 0,
                )
        }
    }

    data class QsoLogged(
        val timeOff: DateTime,
        val dxCall: String,
        val dxGrid: String,
        val txFreqHz: Long,
        val mode: String,
        val reportSent: String,
        val reportReceived: String,
        val txPower: String,
        val comments: String,
        val name: String,
        val timeOn: DateTime,
        val operatorCall: String,
        val myCall: String,
        val myGrid: String,
        val exchangeSent: String,
        val exchangeReceived: String,
        val propMode: String,
    )

    // --- writer -------------------------------------------------------------

    private class Writer {
        val bytes = ByteArrayOutputStream(64)
        private val out = DataOutputStream(bytes)
        fun u8(v: Int) = out.writeByte(v)
        fun bool(v: Boolean) = out.writeByte(if (v) 1 else 0)
        fun i32(v: Int) = out.writeInt(v)
        fun i64(v: Long) = out.writeLong(v)
        fun f64(v: Double) = out.writeDouble(v)
        fun string(s: String) {
            val b = s.toByteArray(Charsets.UTF_8)
            out.writeInt(b.size)
            out.write(b)
        }
        fun datetime(dt: DateTime) {
            out.writeLong(dt.julianDay)
            out.writeInt(dt.msOfDay)
            out.writeByte(1) // Qt::UTC
        }
        fun toByteArray(): ByteArray { out.flush(); return bytes.toByteArray() }
    }

    private fun begin(type: Int): Writer {
        val w = Writer()
        w.i32(MAGIC)
        w.i32(SCHEMA)
        w.i32(type)
        w.string(CLIENT_ID)
        return w
    }

    // --- outbound builders --------------------------------------------------

    fun heartbeat(version: String, revision: String): ByteArray {
        val w = begin(T_HEARTBEAT)
        w.i32(SCHEMA) // max schema
        w.string(version)
        w.string(revision)
        return w.toByteArray()
    }

    fun status(s: Status): ByteArray {
        val w = begin(T_STATUS)
        w.i64(s.dialFreqHz)
        w.string(s.mode)
        w.string(s.dxCall)
        w.string(s.report)
        w.string(s.txMode)
        w.bool(s.txEnabled)
        w.bool(s.transmitting)
        w.bool(s.decoding)
        w.i32(s.rxDf)
        w.i32(s.txDf)
        w.string(s.deCall)
        w.string(s.deGrid)
        w.string(s.dxGrid)
        w.bool(s.txWatchdog)
        w.string(s.subMode)
        w.bool(s.fastMode)
        w.u8(s.specialOpMode)
        w.i32(s.freqTolerance)
        w.i32(s.trPeriod)
        w.string(s.configName)
        w.string(s.txMessage)
        return w.toByteArray()
    }

    fun decode(d: Decode): ByteArray {
        val w = begin(T_DECODE)
        w.bool(d.isNew)
        w.i32(d.timeMs)
        w.i32(d.snr)
        w.f64(d.deltaTime)
        w.i32(d.deltaFreq)
        w.string(d.mode)
        w.string(d.message)
        w.bool(d.lowConfidence)
        w.bool(d.offAir)
        return w.toByteArray()
    }

    fun clear(): ByteArray = begin(T_CLEAR).toByteArray()

    fun qsoLogged(q: QsoLogged): ByteArray {
        val w = begin(T_QSO_LOGGED)
        w.datetime(q.timeOff)
        w.string(q.dxCall)
        w.string(q.dxGrid)
        w.i64(q.txFreqHz)
        w.string(q.mode)
        w.string(q.reportSent)
        w.string(q.reportReceived)
        w.string(q.txPower)
        w.string(q.comments)
        w.string(q.name)
        w.datetime(q.timeOn)
        w.string(q.operatorCall)
        w.string(q.myCall)
        w.string(q.myGrid)
        w.string(q.exchangeSent)
        w.string(q.exchangeReceived)
        w.string(q.propMode)
        return w.toByteArray()
    }

    fun loggedAdif(adif: String): ByteArray {
        val w = begin(T_LOGGED_ADIF)
        w.string(adif)
        return w.toByteArray()
    }

    // --- inbound ------------------------------------------------------------

    sealed class Inbound {
        /**
         * A Reply request. [deltaFreq] is the audio frequency (Hz) the companion asked us
         * to answer on — the caller should honor it so we key up on the requested tone
         * rather than whatever the current TX frequency happens to be (mirrors the iOS
         * port). It is the WSJT-X message's `df` field; 0 means "unspecified".
         */
        data class Reply(val call: String, val grid: String, val snr: Int, val deltaFreq: Int) :
            Inbound()
        object Replay : Inbound()
        data class HaltTx(val autoOnly: Boolean) : Inbound()
        data class FreeText(val text: String, val send: Boolean) : Inbound()
    }

    /** Bounds-checked big-endian reader; every read returns null past the end. */
    private class Reader(private val b: ByteArray) {
        var pos = 0
        private fun take(n: Int): Int? {
            // Reject a negative length (e.g. a hostile/garbled 32-bit string length) and any
            // n that overruns the buffer. Comparing against the remaining bytes (b.size - pos,
            // always >= 0 since pos only advances through validated takes) avoids the
            // pos + n integer overflow a huge n would cause with a naive `pos + n > b.size`.
            if (n < 0 || n > b.size - pos) return null
            val start = pos; pos += n; return start
        }
        fun u8(): Int? { val i = take(1) ?: return null; return b[i].toInt() and 0xFF }
        fun bool(): Boolean? { val v = u8() ?: return null; return v != 0 }
        fun i32(): Int? {
            val i = take(4) ?: return null
            return ((b[i].toInt() and 0xFF) shl 24) or ((b[i + 1].toInt() and 0xFF) shl 16) or
                ((b[i + 2].toInt() and 0xFF) shl 8) or (b[i + 3].toInt() and 0xFF)
        }
        fun f64(): Double? {
            val i = take(8) ?: return null
            var bits = 0L
            for (k in 0 until 8) bits = (bits shl 8) or (b[i + k].toLong() and 0xFF)
            return Double.fromBits(bits)
        }
        fun string(): String? {
            val len = i32() ?: return null
            if (len == -1) return "" // 0xFFFFFFFF null
            val i = take(len) ?: return null
            return String(b, i, len, Charsets.UTF_8)
        }
    }

    /**
     * Parse an inbound datagram into intent. Returns null for anything we don't
     * act on (bad magic, unknown/outbound-only type, truncated buffer).
     */
    fun parseInbound(buf: ByteArray): Inbound? {
        val r = Reader(buf)
        if (r.i32() != MAGIC) return null
        r.i32() ?: return null // schema
        val type = r.i32() ?: return null
        r.string() ?: return null // id
        return when (type) {
            T_REPLY -> {
                r.i32() ?: return null // time
                val snr = r.i32() ?: return null
                r.f64() ?: return null // dt
                val deltaFreq = r.i32() ?: return null // df: the requested audio freq (Hz)
                r.string() ?: return null // mode
                val message = r.string() ?: return null
                val target = parseReplyTarget(message) ?: return null
                Inbound.Reply(target.first, target.second, snr, deltaFreq)
            }
            T_REPLAY -> Inbound.Replay
            // Reject a truncated packet rather than defaulting the trailing bool: a malformed
            // datagram must not trigger an unintended halt or transmit when accepting requests.
            T_HALT_TX -> Inbound.HaltTx(r.bool() ?: return null)
            T_FREE_TEXT -> {
                val text = r.string() ?: return null
                Inbound.FreeText(text, r.bool() ?: return null)
            }
            else -> null
        }
    }

    /**
     * From a decoded message line, work out which station a Reply should call and
     * their grid. Handles `TO FROM ...` and CQ lines with an optional directive
     * (`CQ FROM GRID`, `CQ DX FROM GRID`, `CQ POTA FROM GRID`).
     */
    @VisibleForTesting
    internal fun parseReplyTarget(message: String): Pair<String, String>? {
        val tokens = message.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }
        if (tokens.isEmpty()) return null
        val callIdx = if (tokens[0].equals("CQ", ignoreCase = true)) {
            when {
                tokens.size < 2 -> return null
                isCallsign(tokens[1]) -> 1
                else -> 2
            }
        } else {
            1
        }
        val call = tokens.getOrNull(callIdx) ?: return null
        if (!isCallsign(call)) return null
        val grid = tokens.drop(callIdx + 1).firstOrNull { isGrid4(it) }?.uppercase() ?: ""
        return Pair(call.uppercase(), grid)
    }

    private fun isCallsign(t: String): Boolean {
        val core = t.substringBefore('/')
        val hasAlpha = core.any { it in 'A'..'Z' || it in 'a'..'z' }
        val hasDigit = core.any { it.isDigit() }
        val okChars = t.all { it.isLetterOrDigit() || it == '/' }
        return hasAlpha && hasDigit && okChars && !isGrid4(t)
    }

    private fun isGrid4(t: String): Boolean {
        if (t.length != 4) return false
        val a = t[0].uppercaseChar(); val b = t[1].uppercaseChar()
        return a in 'A'..'R' && b in 'A'..'R' && t[2].isDigit() && t[3].isDigit()
    }

    // --- date/time helpers --------------------------------------------------

    /** Julian Day Number for a UTC `YYYYMMDD` date (matches Qt's toJulianDay). */
    @VisibleForTesting
    internal fun julianDayFromYyyymmdd(s: String?): Long? {
        if (s == null || s.length != 8 || !s.all { it.isDigit() }) return null
        return try {
            val year = s.substring(0, 4).toInt()
            val month = s.substring(4, 6).toInt()
            val day = s.substring(6, 8).toInt()
            // epochDay 0 = 1970-01-01 = JDN 2440588.
            java.time.LocalDate.of(year, month, day).toEpochDay() + 2_440_588L
        } catch (e: Exception) {
            null
        }
    }

    /** Milliseconds since midnight for a UTC `HHMMSS` string. */
    @VisibleForTesting
    internal fun msOfDayFromHhmmss(s: String?): Int? {
        if (s == null || s.length != 6 || !s.all { it.isDigit() }) return null
        val h = s.substring(0, 2).toInt()
        val m = s.substring(2, 4).toInt()
        val sec = s.substring(4, 6).toInt()
        if (h > 23 || m > 59 || sec > 60) return null
        return ((h * 3600) + (m * 60) + sec) * 1000
    }
}
